contracts/Otoken.sol

```diff
@@ -1,12 +1,242 @@
-/**
- * SPDX-License-Identifier: UNLICENSED
- */
-pragma solidity 0.6.10;
+/* SPDX-License-Identifier: UNLICENSED */
+pragma solidity =0.6.10;
+
+import {ERC20Initializable} from "./packages/oz/upgradeability/ERC20Initializable.sol";
+import {SafeMath} from "./packages/oz/SafeMath.sol";
+import {Strings} from "./packages/oz/Strings.sol";
+import {BokkyPooBahsDateTimeLibrary} from "./packages/BokkyPooBahsDateTimeLibrary.sol";
+import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
 
 /**
- *
+ * @title Otoken
+ * @author Opyn
+ * @notice Otoken is the ERC20 token for an option.
+ * @dev The Otoken inherits ERC20Initializable because we need to use the init instead of constructor.
  */
-// solhint-disable-next-line no-empty-blocks
-contract Otoken {
+contract Otoken is ERC20Initializable {
+    using SafeMath for uint256;
+
+    /// @notice address of the addressBook module
+    address public addressBook;
+
+    /// @notice asset that the option references
+    address public underlyingAsset;
+
+    /// @notice asset that the strike price is denominated in
+    address public strikeAsset;
+
+    /// @notice asset that is held as collateral against short/written options
+    address public collateralAsset;
+
+    /// @notice strike price with decimals = 18
+    uint256 public strikePrice;
+
+    /// @notice time of the option represented by unix timestamp
+    uint256 public expiry;
+
+    /// @notice is this a put option, if not it is a call
+    bool public isPut;
+
+    uint256 private constant STRIKE_PRICE_DIGITS = 1e18;
+
+    constructor(address _addressBook) public {
+        addressBook = _addressBook;
+    }
+
+    /**
+     * @notice initialize the otoken.
+     * @param _underlyingAsset asset that the option references
+     * @param _strikeAsset asset that the strike price is denominated in
+     * @param _collateralAsset asset that is held as collateral against short/written options
+     * @param _strikePrice strike price with decimals = 18
+     * @param _expiry time of the option represented by unix timestamp
+     * @param _isPut is this a put option, if not it is a call
+     */
+    function init(
+        address _underlyingAsset,
+        address _strikeAsset,
+        address _collateralAsset,
+        uint256 _strikePrice,
+        uint256 _expiry,
+        bool _isPut
+    ) external initializer {
+        underlyingAsset = _underlyingAsset;
+        strikeAsset = _strikeAsset;
+        collateralAsset = _collateralAsset;
+        strikePrice = _strikePrice;
+        expiry = _expiry;
+        isPut = _isPut;
+        (string memory name, string memory symbol) = _getNameAndSymbol();
+        __ERC20_init_unchained(name, symbol);
```

Comment:
where are the decimals set?

Reply:
is set in the `__ERC20_init_unchained ` function

```javascript
function __ERC20_init_unchained(string memory name, string memory symbol) internal initializer {
        _name = name;
        _symbol = symbol;
        _decimals = 18;
    }
```